primitives/runtime/src/vec_db.rs

```diff
@@ -0,0 +1,274 @@
+use frame_support::{PalletError, StateVersion};
+use sp_core::{storage::TrackedStorageKey, RuntimeDebug};
+use sp_runtime::SaturatedConversion;
+use sp_std::{default::Default, vec, vec::Vec};
+use sp_trie::{
+	generate_trie_proof, verify_trie_proof, LayoutV0, LayoutV1, StorageProof, TrieDBBuilder,
+	TrieHash,
+};
+
+use codec::{Decode, Encode};
+use hash_db::Hasher;
+use scale_info::TypeInfo;
+use trie_db::{DBValue, Trie};
+
+use crate::{storage_proof::RawStorageProof, Size};
+
+pub type RawStorageKey = Vec<u8>;
+
+/// Errors that can occur when interacting with `UntrustedVecDb` and `TrustedVecDb`.
+#[derive(Clone, Encode, Decode, RuntimeDebug, PartialEq, Eq, PalletError, TypeInfo)]
+pub enum VecDbError {
+	UnableToGenerateTrieProof,
+	InvalidProof,
+	UnsortedEntries,
+	UnavailableKey,
+	EmptyVal,
+	DecodeError,
+	UnusedKey,
+}
+
+/// Structure representing a key-value database stored as a sorted `Vec` of tuples.
+///
+/// The structure also contains a proof of the fact that the key-value tuples are actually present
+/// in the chain storage.
+#[derive(Clone, Default, Decode, Encode, Eq, PartialEq, RuntimeDebug, TypeInfo)]
+pub struct UntrustedVecDb {
+	proof: RawStorageProof,
+	db: Vec<(RawStorageKey, Option<DBValue>)>,
+}
+
+impl UntrustedVecDb {
+	/// Creates a new instance of `UntrustedVecDb`.
+	pub fn new<H: Hasher>(
+		read_proof: StorageProof,
+		root: TrieHash<LayoutV1<H>>,
+		mut keys: Vec<impl AsRef<[u8]> + Ord>,
+	) -> Result<Self, VecDbError> {
+		let mem_db = read_proof.into_memory_db();
+		let trie_db = TrieDBBuilder::<LayoutV1<H>>::new(&mem_db, &root).build();
```

Comment:
Sorry, I still had no time to explore this layout stuff in details, so probably an absolute dumb q - is it fine that we're always using `LayoutV1` here, but switching between `LayoutV0` and `LayoutV1` when verifying proofs?